Source/santactl/Commands/SNTCommandEventUpload.mm

```diff
@@ -0,0 +1,218 @@
+#import "Source/common/SNTXPCBundleServiceInterface.h"
+#import "Source/common/SNTXPCControlInterface.h"
+#import "Source/common/SNTXPCSyncServiceInterface.h"
+#import "Source/common/SigningIDHelpers.h"
+#import "Source/common/StoredEventHelpers.h"
+#import "Source/santactl/SNTCommand.h"
+#import "Source/santactl/SNTCommandController.h"
+
+@interface SNTCommandEventUpload : SNTCommand <SNTCommandProtocol, SNTBundleServiceProgressXPC>
+@property(atomic) uint64_t currentBinaryCount;
+@property(atomic) uint64_t currentFileCount;
+@property(atomic) uint64_t currentHashedCount;
+@end
+
+@implementation SNTCommandEventUpload
+
+REGISTER_COMMAND_NAME(@"eventupload")
+
++ (BOOL)requiresRoot {
+  return NO;
+}
+
++ (BOOL)requiresDaemonConn {
+  return YES;
+}
+
++ (NSString *)shortHelpText {
+  return @"Generate and upload non-execution events to the sync server.";
+}
+
++ (NSString *)longHelpText {
+  return @"Generates a non-execution event containing the metadata of the provided bundle or file."
+         @"\n"
+         @"Usage: santactl eventupload [options] [file-or-bundle-paths]\n"
+         @"    --binary-only: Create a single event for the provided files only. No surrounding\n"
+         @"                   bundle information will be uploaded.\n"
+         @"\n"
+         @"Examples: santactl eventupload /Applications/Google\\ Chrome.app\n"
+         @"          santactl eventupload /usr/bin/yes /usr/bin/cal\n"
+         @"          santactl eventupload --binary-only /Applications/Google\\ "
+         @"Chrome.app/Contents/MacOS/Google\\ Chrome\n"
+         @"\n"
+         @"Note: A sync server must be configured. To generate and upload bundle events, the sync\n"
+         @"server must support bundle hashing.";
+}
+
++ (BOOL)isHidden {
+  return NO;
+}
+
+- (NSArray *)pathsFromArguments:(NSArray *)arguments skipBundles:(BOOL *)skipBundles {
+  NSUInteger noBundleIndex = [arguments indexOfObject:@"--binary-only"];
+  if (noBundleIndex != NSNotFound) {
+    if (skipBundles) *skipBundles = YES;
+    NSMutableArray *newArguments = arguments.mutableCopy;
+    [newArguments removeObjectAtIndex:noBundleIndex];
+    arguments = [newArguments copy];
+  }
+  return arguments;
+}
+
+- (void)runWithArguments:(NSArray *)arguments {
+  if (![SNTConfigurator configurator].syncBaseURL) {
+    LOGE(@"Missing SyncBaseURL. Exiting.");
+    exit(1);
+  }
+
+  // Parse arguments
+  if (!arguments.count) [self printErrorUsageAndExit:@"No arguments"];
+  BOOL skipBundles = NO;
+  NSArray *paths = [self pathsFromArguments:arguments skipBundles:&skipBundles];
+  NSMutableArray *events = [NSMutableArray arrayWithCapacity:paths.count];
+
+  // Hash the bundle if the server declares support and if the caller has not explicitly disabled
+  // bundle hashing.
+  __block BOOL enableBundles = NO;
+  [self.daemonConn.synchronousRemoteObjectProxy enableBundles:^(BOOL response) {
+    enableBundles = response;
+  }];
+  if (skipBundles) {
+    enableBundles = false;
+  }
```

Comment:
Nit - the NO/false double setting took me a bit to parse. Maybe something like this is cleaner, and also let's us skip asking the daemon if not necessary:
```objc
  __block BOOL enableBundles = !skipBundles;
  if (enableBundles) {
    [self.daemonConn.synchronousRemoteObjectProxy enableBundles:^(BOOL response) {
      enableBundles = response;
    }];
  }
```